phoenix-core/src/it/java/org/apache/phoenix/end2end/InvalidateMetadataCacheIT.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.phoenix.thirdparty.com.google.common.collect.Maps;
+import org.apache.phoenix.util.PropertiesUtil;
+import org.apache.phoenix.util.ReadOnlyProps;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.util.Map;
+import java.util.Properties;
+
+import static org.apache.hadoop.hbase.coprocessor.CoprocessorHost.REGIONSERVER_COPROCESSOR_CONF_KEY;
+import static org.apache.phoenix.coprocessor.MetaDataEndpointImpl.PHOENIX_METADATA_CACHE_INVALIDATION_TIMEOUT_MS;
+import static org.apache.phoenix.util.TestUtil.TEST_PROPERTIES;
+import static org.junit.Assert.fail;
+
+@Category({NeedsOwnMiniClusterTest.class })
+public class InvalidateMetadataCacheIT extends BaseTest {
+    private static final Logger LOGGER = LoggerFactory.getLogger(InvalidateMetadataCacheIT.class);
+
+    @BeforeClass
+    public static synchronized void doSetup() throws Exception {
+        NUM_SLAVES_BASE = 2;
+        Map<String, String> props = Maps.newHashMapWithExpectedSize(1);
+        // to fail fast in case of exception.
+        props.put("hbase.client.retries.number", String.valueOf(0));
+        props.put(REGIONSERVER_COPROCESSOR_CONF_KEY,
+                FailingPhoenixRegionServerEndpoint.class.getName());
+        // Setting phoenix metadata cache invalidation timeout to a small number to fail fast.
+        props.put(PHOENIX_METADATA_CACHE_INVALIDATION_TIMEOUT_MS, String.valueOf(2000));
+        setUpTestDriver(new ReadOnlyProps(props.entrySet().iterator()));
+    }
+
+    /**
+     * Add FailingPhoenixRegionServerEndpoint as regionserver co-processor.
+     * Make one of the regionserver sleep in invalidateServerMetadataCache method. This will trigger
+     * TimeoutException in MetadataEndpointImpl#invalidateServerMetadataCacheWithRetries method.
+     * Make sure that ALTER TABLE ADD COLUMN statement fails.
+     */
+    @Test
+    public void testAddColumnWithTimeout() {
+        Properties props = PropertiesUtil.deepCopy(TEST_PROPERTIES);
+        String dataTableFullName = generateUniqueName();
+        String ddl = getCreateTableStmt(dataTableFullName);
+        HRegionServer regionServerZero = utility.getMiniHBaseCluster().getRegionServer(0);
+        FailingPhoenixRegionServerEndpoint coprocForRS0 =
+                getFailingPhoenixRegionServerEndpoint(regionServerZero);
+        coprocForRS0.sleep();
+        try(Connection conn = DriverManager.getConnection(getUrl(), props)) {
+            conn.createStatement().execute(ddl);
+            conn.createStatement().execute("ALTER TABLE " + dataTableFullName +
+                    " ADD CF.col2 integer");
+            fail("Shouldn't reach heere");
+        } catch (Exception e) {
+            LOGGER.error("Exception while adding column", e);
+            // This is expected
```

Comment:
should we add assertTrue in this exception? similarly below?

Reply:
I thought that is kind of implicit. There is an `assert fail` statement in the try block and if it throws anything other than Exception, the test will fail. So it has to go in `catch (Exception)` block for the test to succeed.